### DISCUSSION TITLE
Is there a data sources usage example

### DISCUSSION THREAD

**Discussion by Janik-Haag** [netbox-community/netbox#14495] [Q&A] [answered]
Hi, I was trying to play around with netbox data sources, but can't seem to find a usage example on how to use them. I tried adding one and syncing data from a git repo but it always just throws `null` in the error field of the job and I have read about some cli being supposed to throw more messages but can't seem to find it. Would anyone mind pointing me in the right direction/linking a example or usage guide?

**Comment by llamafilm** [comment]
Have you seen this? https://github.com/netbox-community/netbox/blob/develop/docs/models/core/datasource.md

**Reply by Janik-Haag** [reply]
Yes, but it does not show you what a datasource file should look like.

**Comment by jeffgdotorg** [comment, marked as answer]
What goals are you aiming to accomplish using a data source? There are a few ways to use them.

This [four-minute video](https://www.youtube.com/watch?v=jxsFwyMk18k) shows how to add a Git data source, and then set up and run a script and a report from the data source. Perhaps it will provide some helpful context.

**Reply by Janik-Haag** [reply]
Thank you for sharing that video, that helped a lot.
The first error was just me using the demo instance and thus not having internet access wich just throws the error null..

> What goals are you aiming to accomplish using a data source? 

So I want to store most kinds of virtual resources like ips, vlans, l2vpns, interfaces, services, etc... on the netbox (preferably) filesystem or in a git repo and then sync them every few minutes if something changed since I already have most of that data in a git repository. The reasons for wanting to load it from disk is to be able to run some pre-processing without accumulating more and more commits in some random git repo.

I found how to structure datafiles ^^
The only thing I'm missing now is how to sync a datafile in things like for example the ipam/ip-addresses like I can with Customization/Reports or Customization/Scripts. Just regularly running import's won't work because if I delete an entry from a datafile I also want the entry to be deleted from the database on the next sync.

I really appreciate your help :3

**Reply by zndrr** [reply]
Is there much in the way of guidance for loading private repos?
I did an initial attempt but naturally wasn't met with success. Though admittedly haven't invested much time in to it yet.

FWIW self-hosted Gitlab authed repos run fine. Just struggled with GitHub private

**Reply by jeffgdotorg** [reply]
I think this question merits a separate discussion.
